Apply track_caller to closure on `expect_non_local()`

JohnTitor opened the pull request:
r? @petrochenkov 

Alternatively we could remove the closure by replicating the same logic of `map_id()`. I'm happy to switch to it if you'd like.

JohnTitor commented:
(Will fix the CI failure after discussing a preferred way)

petrochenkov commented:
Using `#[track_caller]` (and also `#[cold]`) on the closure would be the better alternative both here and in https://github.com/rust-lang/rust/pull/96778.

However, something is broken in feature gating of `#[track_caller]` on closures, and they cannot be "ungated" even if the corresponding `#![feature]` is in place.
It would be great to find out why it happens and fix that issue.

JohnTitor commented:
Hmm, I'm afraid that I'm not familiar with it. At least, looking at https://github.com/rust-lang/rust/pull/87064 and related code, feature gating is fine. We use `expect_non_local()` on rustc_resolve and librustdoc, I think this should be related (actually local build passes when I added the attr on rustc_resolve along with rustc_hir), right? My guess is an attribute (`track_caller`, in this case) is also added at callsite then a crate containing that callsite will need a feature attr.
However I don't think I can resolve this issue only by my hand, any help would be appreciated.

petrochenkov commented:
cc @Aaron1011

Aaron1011 commented:
This looks like a broader issue - the `codegen_fn_attrs` query does feature-gating, and used to be called for foreign `DefIds`. This was recently changed in https://github.com/rust-lang/rust/pull/96473 - and I confirmed that this issue doesn't occur with `./x.py build --stage 2 src/tools/rustdoc`.

For now, I think you could just add `#![feature(closure_track_caller)]` to rustdoc behind a `cfg_attr(not(bootstrap))`, and let someone remove it after the next beta bump.

JohnTitor commented:
👍, how about @petrochenkov?

JohnTitor commented:
By the way, is there an issue tracking that weird behavior?

JohnTitor commented:
Added the feature attribute to librustdoc as suggested, @petrochenkov could you take a look?

petrochenkov commented:
https://github.com/rust-lang/rust/pull/97214 has been merged.
@rustbot author

petrochenkov commented:
Thanks!
@bors r+ rollup

bors commented:
:pushpin: Commit c7db4b08fff9591adb2c7a66d5f81ce685655622 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov c7db4b08fff9591adb2c7a66d5f81ce685655622 -->
<!-- homu: {"type":"Approved","sha":"c7db4b08fff9591adb2c7a66d5f81ce685655622","approver":"petrochenkov"} -->
